File: frontend/src/features/tasks/types/task.ts
import type { Category } from '../../categories/types/category';
import type { Tag } from '../../tags/types/tag';

export type Priority = 'high' | 'medium' | 'low';

export interface Task {
  id: number;
  title: string;
  description: string | null;
  category_id: number | null;
  priority: Priority;
  due_date: string | null;
  is_completed: boolean;
  user_id: number;
  created_at: string;
  updated_at: string;
  tags: Tag[];
  category: Category | null;
}

export interface TaskFilters {
  category_id?: number;
  tag_id?: number;
  is_completed?: boolean;
  page?: number;
  page_size?: number;
}

export interface TaskListResponse {
  total: number;
  page: number;
  page_size: number;
  tasks: Task[];
}

export interface CreateTaskInput {
  title: string;
  description?: string;
  category_id?: number;
  priority?: Priority;
  due_date?: string;
  tag_ids?: number[];
}

export interface UpdateTaskInput {
  title?: string;
  description?: string | null;
  category_id?: number | null;
  priority?: Priority;
  due_date?: string | null;
  tag_ids?: number[];
}

File: backend/app/schemas/task.py
from pydantic import BaseModel, ConfigDict, Field, field_validator
from datetime import datetime
from typing import Optional, List
from app.models.enums import Priority


class TaskCreate(BaseModel):
    title: str = Field(..., min_length=1)
    description: Optional[str] = None
    category_id: Optional[int] = None
    priority: Optional[Priority] = None
    due_date: Optional[datetime] = None
    tag_ids: Optional[List[int]] = []


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    category_id: Optional[int] = None
    priority: Optional[Priority] = None
    due_date: Optional[datetime] = None
    is_completed: Optional[bool] = None
    tag_ids: Optional[List[int]] = None


class TaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: Optional[str]
    category_id: Optional[int]
    priority: Optional[Priority]
    due_date: Optional[datetime]
    is_completed: bool
    user_id: int
    created_at: datetime
    updated_at: datetime


class TaskListResponse(BaseModel):
    tasks: List[TaskResponse]
    total: int
    page: int
    page_size: int
